src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs

```diff
@@ -104,35 +62,57 @@ private void OnConnectionChanged(bool connected)
             ConnectionChanged?.Invoke(this, connected);
         }
 
-        private async Task<PinnedRemotableDataScope> GetPinnedScopeAsync(Solution solution, CancellationToken cancellationToken)
+        /// <summary>
+        /// NoOpClient is used if a user killed our remote host process. Basically this client never
+        /// create a session
+        /// </summary>
+        public class NoOpClient : RemoteHostClient
         {
-            if (solution == null)
+            public NoOpClient(Workspace workspace) :
+                base(workspace)
+            {
+            }
+
+            public override Task<Session> TryCreateSessionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
             {
-                return null;
+                return SpecializedTasks.Default<Session>();
             }
 
-            Contract.ThrowIfFalse(solution.Workspace == _workspace);
+            protected override void OnConnected()
+            {
+                // do nothing
+            }
 
-            var service = _workspace.Services.GetService<ISolutionSynchronizationService>();
-            return await service.CreatePinnedRemotableDataScopeAsync(solution, cancellationToken).ConfigureAwait(false);
+            protected override void OnDisconnected()
+            {
+                // do nothing
+            }
         }
 
-        // TODO: make this to not exposed to caller. abstract all of these under Request and Response mechanism
         public abstract class Session : IDisposable
         {
-            protected readonly PinnedRemotableDataScope PinnedScopeOpt;
             protected readonly CancellationToken CancellationToken;
+            private PinnedRemotableDataScope _scope;
 
             private bool _disposed;
 
-            protected Session(PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+            protected Session(CancellationToken cancellationToken)
             {
                 _disposed = false;
+                _scope = null;
 
-                PinnedScopeOpt = scope;
                 CancellationToken = cancellationToken;
             }
 
+            public PinnedRemotableDataScope PinnedRemotableDataScope => _scope;
```

Comment:
> so are you saying that API should be "internal"?

Pretty much.  I'm saying: think of the Features layer as being something would consume your 'public' API.  So make the public part of your API as simple and clear as possible, without exposing internal details that aren't necessary.  If possible, actually use interfaces for this so that the feature layers are not exposed to internal implementation details like this.

--

In other words, imagine you were making a public API, even if it is going to only be 'internal' for now.